cpp/src/io/csv/writer_impl.cu

```diff
@@ -0,0 +1,545 @@
+    string_view d_str        = d_column_.template element<string_view>(idx);
+    size_type str_size_bytes = d_str.size_bytes();
+
+    char* d_buffer = get_output_ptr(idx);
+    // assert( d_buffer != nullptr );
+
+    if (predicate_(d_str)) {
+      char const quote_char   = '\"';
+      char const* quote_str   = "\"";
+      char const* str_2quotes = "\"\"";
+
+      size_type len1quote{1};
+      size_type len2quotes{2};
+
+      // modify d_str by duplicating all 2bl quotes
+      // and surrounding whole string by 2bl quotes:
+      //
+      // pre-condition: `d_str` is _not_ modified by `d_buffer` manipulation
+      // because it's a copy of `idx` entry in `d_column_`
+      //(since `d_column` is const)
+      //
+      d_buffer = copy_and_increment(d_buffer, quote_str, len1quote);  // add the quote prefix
+
+      for (auto itr = d_str.begin(); itr != d_str.end(); ++itr) {
+        char_utf8 the_chr = *itr;
+
+        if (the_chr == quote_char) {
+          d_buffer = copy_and_increment(d_buffer, str_2quotes, len2quotes);  // double the quote;
+        } else {
+          d_buffer += from_char_utf8(the_chr, d_buffer);
+        }
+      }
+
+      d_buffer = copy_and_increment(d_buffer, quote_str, len1quote);  // add the quote suffix;
+    } else {
+      // copy the source string unmodified:
+      //(pass-through)
+      //
+      memcpy(d_chars_, d_str.data(), str_size_bytes);
+    }
+    return 0;
+  }
+
+  __device__ char* get_output_ptr(size_type idx)
+  {
+    return d_chars_ && d_offsets_ ? d_chars_ + d_offsets_[idx] : nullptr;
+  }
+
+ private:
+  column_device_view const d_column_;
+  int32_t const* d_offsets_;
+  char* d_chars_;
+  predicate_special_chars predicate_;
+};
+
+struct column_to_strings_fn {
+  // compile-time predicate that defines unsupported column types;
+  // based on the conditions used for instantiations of individual
+  // converters in strings/convert/convert_*.hpp;
+  //(this should have been a `variable template`,
+  // instead of a static function, but nvcc (10.0)
+  // fails to compile var-templs);
+  //
+  template <typename column_type>
+  constexpr static bool is_not_handled(void)
+  {
+    return (  //(!std::is_same<column_type, bool>::value) && <- case covered by is_integral
+      (!std::is_same<column_type, cudf::string_view>::value) &&
+      (!std::is_integral<column_type>::value) && (!std::is_floating_point<column_type>::value) &&
+      (!cudf::is_timestamp<column_type>()));
+  }
+
+  explicit column_to_strings_fn(writer_options const& options,
+                                rmm::mr::device_memory_resource* mr = nullptr,
+                                cudaStream_t stream                 = nullptr)
+    : options_(options), mr_(mr), stream_(stream)
+  {
+  }
+
+  // Note: Done: `null` replacement with `na_rep` defered to `concatenate()`
+  // instead of column-wise; might be faster
+  //
+  // Note: Cannot pass `stream` to detail::<fname> version of <fname> calls below, because they are
```

Comment:
The right thing is to create `include/cudf/strings/detail/` headers for these but we could do that in a separate PR if you want.